general/mdspan.hpp

```diff
@@ -0,0 +1,417 @@
+// Copyright (c) 2010-2023, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_MDSPAN_HPP
+#define MFEM_MDSPAN_HPP
+
+#include <list>
+#include <array>
+#include <vector>
+#include <utility>
+#include <type_traits>
+
+#include "device.hpp"
+#include "backends.hpp"
+
+namespace mfem
+{
+
+namespace internal // experimental helper functions for mfem::MDLayout
+{
+
+// md_sequence represents a compile-time sequence of integers
+template <typename T, T... args> struct md_sequence { };
+
+template <typename T, int N, bool left> struct make_md_sequence;
+
+// make_sequence, specialized for left (default) and right layout
+template <typename T, int N, bool left = true>
+using make_sequence = typename make_md_sequence<T, N, left>::type;
+
+} // namespace internal
+
+/// @brief The MDOffset class computes the multi-dimensional offsets
+template <int n, int N, typename T, typename... Ts>
+struct MDOffset
+{
+   static MFEM_HOST_DEVICE inline
+   T offset(const int (&Sd)[N], T nd, Ts... args)
+   { return nd * Sd[n-1] + MDOffset<n+1, N, Ts...>::offset(Sd, args...); }
+};
+
+template <int N, typename T, typename... Ts>
+struct MDOffset<N, N, T, Ts...>
+{
+   static MFEM_HOST_DEVICE inline
+   T offset(const int (&Sd)[N], T nd) { return nd * Sd[N-1]; }
+};
+
+/// @brief The MDTensor class holds the pointer and strides for each dimension
+template<int N, typename T> class MDTensor
+{
+   T *ptr;
+   int Sd[N];
+
+public:
+   /// Default constructor
+   MDTensor() = delete;
+
+   /// Copy constructor (default)
+   MDTensor(const MDTensor&) = default;
+
+   /// Copy assignment (default)
+   MDTensor& operator=(const MDTensor&) = default;
+
+   /// Constructor to initialize a tensor from a pointer and strides
+   template <typename... Args> MFEM_HOST_DEVICE
+   MDTensor(T *ptr, const int (&sd)[N]): ptr(ptr)
+   { for (int i = 0; i < N; ++i) { Sd[i] = sd[i]; } }
+
+   /// Accessor for the data
+   template <typename... Ts> MFEM_HOST_DEVICE inline
+   T& operator()(Ts... args) { return ptr[Offset(args...)]; }
```

Comment:
This version of `operator()` can be removed -- the const version below does exactly the same.